### ISSUE TITLE
Server responses with HTTP 408 with big indexes

### ISSUE THREAD

**Issue by bumzack** [meilisearch/meilisearch#3653]
**Describe the bug**
I have a meilisearch instance with 3 indexes. I am reading documents from one index, then run filtered queries for each document on the other two indexes. I combine the data from the 3 documents and write these new documents to a fourth index.
Immediately after I start the program, the meilisearch rust client logs warnings with a 408 response code: 
 
```WARN  meilisearch_sdk::request] Expected response code 200, got 408```

**To Reproduce**
I've put together a repository with code to reproduce the problem: https://github.com/bumzack/meili_reproducible_408  

Basically this code creates 3 indexes with dummy documents: one with  50_000_000 documents and two with 10_000_000 documents. These are the orders of magnitude from the original dataset I am using and where i first encountered this problem https://datasets.imdbws.com 
The filter requests always return the same amount of documents for the indexes (2 resp 6 documents).

Running the binary from the repo, the first 408 response appears almost immediately after the start of the program.  
The app spawns 4 tokio tasks and each task reads documents paginated with a limit of 100. 

The problem seems to be coupled to the size of the indexes. I tried smaller index sizes and increased the number of tasks, which did not trigger the 408 responses. 

Log from the program in the repo
```
[2023-04-08T18:51:50Z WARN  meilisearch_sdk::request] Expected response code 200, got 408
[2023-04-08T18:51:50Z INFO  create_combined_index] found 100 movies
[2023-04-08T18:51:50Z INFO  create_combined_index] error reading persons Error parsing response JSON: invalid type: null, expected struct MeilisearchError at line 1 column 4
thread 'tokio-runtime-worker' panicked at 'should not happen. error reading persons.', create_combined_index/src/main.rs:142:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-04-08T18:51:50Z WARN  meilisearch_sdk::request] Expected response code 200, got 408
[2023-04-08T18:51:50Z INFO  create_combined_index] error reading persons Error parsing response JSON: invalid type: null, expected struct MeilisearchError at line 1 column 4
thread 'tokio-runtime-worker' panicked at 'should not happen. error reading persons.', create_combined_index/src/main.rs:142:13
[2023-04-08T18:51:50Z WARN  meilisearch_sdk::request] Expected response code 200, got 408
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: ParseError(Error("invalid type: null, expected struct MeilisearchError", line: 1, column: 4))', create_combined_index/src/main.rs:118:10
```

**Expected behavior**
I would not expect any 408 responses from the server. The server load with 4 tasks and no other clients querying the server seem like a reasonable workload to me. 

**Screenshots**
n/a

**Meilisearch version:**
v1.1.0
rustc: 1.68.2

**Additional context**
The meilisearch instance is cloned from the github repo. 
Tag v1.1.0 checked out. 
Compiled in release mode. 
```config.toml``` can be found in the repo readme
  
OS: Ubunutu 22
RAM: 128 GB
12 cores

The same problem was already reported in #1202 and closed as not reproducible. Thats why I put together the not so minimalistic  MRE. 

If you need additional information please let me know.

**Comment by loiclec** [comment]
Hi @bumzack ,

What a fantastic bug report, thank you so much! 
We will reproduce it using the code you provided and come back to you when we found the root cause or if we have more questions. Unfortunately, you may have to wait ~2 weeks to get a good reply, because a lot of the company will be unavailable next week. Nevertheless, it remains a priority for us, especially after all the care you took to surface this problem to us 🙏

**Comment by bumzack** [comment]
Hi @loiclec 

thank your for your kind words :-)

I just wanted to add one piece of information I forgot: the meilisearch instance is running behind a nginx server. So the client(s) from the repo do not talk directly to meilisearch server. In case it makes any difference.

Thanks for looking into the problem. I am evaluating meilisearch and Solr and I am trying to use Rust all the way :-)

I am happy to provide additional information or test cases if needed. Please don't hesitate to ask.

All the best!

**Comment by dureuill** [comment]
Hello @bumzack :wave:,

I tried reproducing the issue from your reproducer on a mac M1, but did not reproduce.

Thank you for the additional information. It is possible that the HTTP 408 is due to the nginx proxy being configured to close the connection if the Meilisearch server is too long to respond (due to being under heavy load).

Can you check the below items in your `nginx.conf`:
    - `client_body_timeout`
    - `client_header_timeout`
    - `keepalive_timeout`
Does increasing the value of these configuration items eliminate the HTTP 408 errors?

**Comment by dgsharpe** [comment]
I am experiencing this same issue on large indexes, but am not using Nginx.

**Comment by dureuill** [comment]
Hello @dgsharpe :wave:

Thank you for taking the time to report this issue as well. Your setup prompted me to try and reproduce this again, and this time I was successful in reproducing the issue by artificially blocking the worker threads that are handling requests. 

I found out about https://github.com/actix/actix-web/issues/1269, which seems directly related to the issue at hand.

I made a reproducer using only actix-web rather than Meilisearch. It can be found here: https://github.com/dureuill/issue-3653-reproducer.

I will make a post in the issue discussion I linked above. My naive understanding of the expected behavior is that in the best case, connections should be accepted and kept as "pending" when no workers are available, and in the worst case of all workers being exhausted for way too long, then a HTTP 503 should be returned rather than an HTTP 408.

> Please let me know what information I could provide to help debug.

It is possible that Meilisearch blocks for too long on a worker thread for some requests, which it should not. Are you seeing specific requests that take a long time to complete? How many concurrent requests is the server typically handling when the error occurs? Any help is appreciated :heart:

**Comment by dgsharpe** [comment]
@dureuill I'm best able to reproduce this issue when I'm rebuilding one of my larger stores, by which I mean, regenerating every single document from scratch and shoving them into Meilisearch as fast as possible. The documents are of modest size, containing 16 fields, which are ints or booleans or very short text fields (most are just 32 characters, one is ~80 characters). There are about 800k documents in total, and I'm sending them to Meilisearch in batches of 10k. There's several seconds' delay between batches. It's usually one of these requests that fails, rather than the other unrelated requests which are occurring during this rebuild.

The other traffic coming in at this time maxes out at about 60 requests per minute, usually somewhat less. It's bursty, sometimes ~10 requests come in at once. Mostly reads. The server is a small EC2 instance with only two cores. I've verified that the server is not running out of RAM, CPU credits, storage IOPS or bandwidth, etc, etc.

**Comment by dureuill** [comment]
Thank you for the additional information, @dgsharpe!

Having two cores means that only 2 actix workers are spawned, so if the load is such that one client request has to wait more than 5 seconds before a worker can handle it (because it is busy with previous indexing/search requests), then the 408 is returned.

While we're waiting on the upstream issue, maybe we could apply some workaround like raising the minimum number of workers (such that even if a task does a lot of synchronous work by mistake, the OS schedules a different worker thread to advance another task), or raising (or disabling?) this client timeout (with the downside that we will keep alive clients that actually don't send the full query fast enough for longer than necessary)?

**Comment by dureuill** [comment]
Ping from triage

- Severity: 
  - Error should be an HTTP 503 rather than an HTTP 408
  - Error should be mitigated by increasing concurrency for machines that have low parallelism
  - Workarounds:
    - Teach clients to handle 408 as 503 😞 
    - Use a higher spec machine, or decrease search parallelism
 - Size: Expecting large, because it may involve patches to actix. A small mitigation could be try increasing concurrency when parallelism is low (create minimum 8 actix workers even if the machine only has 2 physical cores)

**Comment by bumzack** [comment]
> Hello @bumzack 👋,
> 
> I tried reproducing the issue from your reproducer on a mac M1, but did not reproduce.
> 
> Thank you for the additional information. It is possible that the HTTP 408 is due to the nginx proxy being configured to close the connection if the Meilisearch server is too long to respond (due to being under heavy load).
> 
> Can you check the below items in your `nginx.conf`: - `client_body_timeout` - `client_header_timeout` - `keepalive_timeout` Does increasing the value of these configuration items eliminate the HTTP 408 errors?

sorry for never responding, thanx for alle the effort you put into investigating this issue!! 
much appreciated.

**Comment by IvanVlasic** [comment]
@dureuill hello, is there any updates to this?

We seem to be experiencing same issue. For very short periods of times, less than a second, meilisearch returns 408 for all requests and then goes back to normal. I am positive we are not hitting any timeouts since it seems to be almost instant.

We are using 16 cores which are idling most of the time but I assume it can happen that for a very brief period of time all of them are doing something which causes issue from OP, all threads are busy and instead of waiting for one of them 408 response is returned immediately.

Using higher spec machine is not an option since we're already over provisioned, our CPU runs at about 5% most of the time. We are way below our limits for disk IOPS and don't have any issues or latency on that side, i/o wait seems to be non existent as well.

Do we have any options here?

**Comment by dureuill** [comment]
Hello @IvanVlasic :wave:

No update from our side at the moment, unfortunately. 

Looking at the source the only place that appears to be returning a 408 is the `client_request_timeout` of actix: https://github.com/actix/actix-web/blob/3acdda48e0e5440d426cc81602201b2f1104662f/actix-http/src/h1/dispatcher.rs#L839

An easy way to test this hypothesis would be to recompile Meilisearch configuring the [`client_request_timeout`](https://docs.rs/actix-web/latest/actix_web/struct.HttpServer.html#method.client_request_timeout) to 0, so as to disable it, and see if the error disappears.

You could test this by applying the following patch: 

```diff
diff --git a/meilisearch/src/main.rs b/meilisearch/src/main.rs
index a3905d451..6798ef070 100644
--- a/meilisearch/src/main.rs
+++ b/meilisearch/src/main.rs
@@ -104,7 +104,8 @@ async fn run_http(
     })
     // Disable signals allows the server to terminate immediately when a user enter CTRL-C
     .disable_signals()
-    .keep_alive(KeepAlive::Os);
+    .keep_alive(KeepAlive::Os)
+    .client_request_timeout(std::time::Duration::ZERO);

     if let Some(config) = opt_clone.get_ssl_config()? {
         http_server.bind_rustls(opt_clone.http_addr, config)?.run().await?;
```

and then recompiling.

If it is the timeout, then it is caused by https://github.com/actix/actix-web/issues/1269

Sorry I can't be of more help, it is hard to work on issues we don't locally reproduce.

**Comment by devoNOTbevo** [comment]
FYI, this is an issue that we started having an issue with Meili returning 408's. We use Laravel Scout and have a big text field we sync over each document (ocr representation of a PDF). We tried doing smaller batches to sync at first, but still happened.

We went up from T2 Small to T2 Medium (from 1 to 2 cores) after reading the actix thread, thinking that this would at least let search queries and indexing happen without stepping on one another.

So far we haven't seen a single 408 since we did that.

So for posterity - if you are:

1. running on 1 to 2 cores.
2. see 408 exceptions.
3. but don't see any spikes in connections, CPU capacity, or RAM that explain a timeout.

then it may be how this, which boils down to how actix is handling responding to the requests.
